src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,310 @@
+        copy and overhead of atomic per chunk copy. Best chunk size should
+        balance these two factors.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_do_lmt_coop_copy(const void *src_buf,
+                                                          size_t data_sz, void *dest_buf,
+                                                          OPA_int_t * counter_ptr,
+                                                          uint64_t req_ptr,
+                                                          MPIR_Request * local_req, int *fin_type)
 {
     int mpi_errno = MPI_SUCCESS;
+    size_t copy_sz;
+    int cur_chunk = 0, total_chunk = 0, num_local_copy = 0;
+    uint64_t cur_offset;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
+
+    total_chunk =
+        data_sz / MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE +
+        (data_sz % MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE == 0 ? 0 : 1);
+
+    MPIR_Assert(total_chunk > 1);
+
+    /* TODO: implement OPA_fetch_and_incr_int uint64_t to directly change cur_offset
+     * instead of computing it in each copy */
+    while ((cur_chunk = OPA_fetch_and_incr_int(counter_ptr)) < total_chunk) {
+        cur_offset = ((uint64_t) cur_chunk) * MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE;
+        copy_sz =
+            cur_offset + MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE <=
+            data_sz ? MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE : data_sz - cur_offset;
+        mpi_errno =
+            MPIR_Localcopy(((char *) src_buf + cur_offset), copy_sz, MPI_BYTE,
+                           ((char *) dest_buf + cur_offset), copy_sz, MPI_BYTE);
+        MPIR_ERR_CHECK(mpi_errno);
+        num_local_copy++;
+    }
+
+    /* Note:
+     * If receiver copies all chunks, it shouldn't wait sender's fin packet. Instead,
+     * receiver checks its number of local copy and send fin packet to sender if
+     * num_local_copy == total_chunk.
+     * When sender and receiver both copy a part of chunks, only who gets
+     * cur_chunk == (total_chunk + 1) will send fin packet to the other side.
+     * For example:
+     * If sender is copying last chunk, its cur_chunk should be total_chunk - 1,
+     * then if receiver gets cur_chunk == total_chunk, it means there is no chunk
+     * to be copied but copy may be still under processing, so receiver must exit this
+     * routine and wait for fin packet from sender.
+     * finally sender gets cur_chunk == (total_chunk + 1) which means copy is complete
+     * and sends fin_type packet to receiver. */
+    if (num_local_copy == total_chunk)
+        *fin_type = MPIDI_XPMEM_COPY_ALL_ISSUE_FIN;
+    else if (num_local_copy == 0)
+        *fin_type = MPIDI_XPMEM_WAIT_FIN;
+    else if (cur_chunk == total_chunk + 1)
+        *fin_type = MPIDI_XPMEM_COPY_MIX_ISSUE_FIN;
+    else
+        *fin_type = MPIDI_XPMEM_WAIT_FIN;
+
```

Comment:
The `fin_type` values look overcomplicated to me. I think we only need a bool type:
- `copy_fin==TRUE`: 
   `num_local_copy == total_chunk` or `num_local_copy != total_chunk && cur_chunk == total_chunk + 1`. 
    The caller will issue the FIN packet to the peer. If it is receiver, it also releases the counter.
- `copy_fin==FALSE`:
    The caller does nothing

When the `FIN` packet is received, if it is sender it will deregist it from the tree; if it is receiver it will release the counter.

Is the above logic correct?

Reply:
I need the third type to distinguish `num_local_copy == total_chunk` and `cur_chunk == total_chunk + 1` cases. When `num_local_copy == total_chunk`, receiver cannot release counter right away and has to enqueue it for future release; under `cur_chunk == total_chunk + 1` case, receiver can release counter right away.

Reply:
I think we might be able to avoid free counter queue. Here is my thought:
- define two `fin_type`:
     `LOCAL_FIN`: locally finished (`cur_chunk==total_chunk`)
     `ALL_FIN`: both sides finished (`cur_chunk==total_chunk+1`)
- define three `copy_stat`:
      `LOCAL_COPIED`: `num_local_copied==total_chunk`
      `REMOTE_COPIED`: `num_local_copied==0`
      `MIXED_COPIED`: `0<num_local_copied<total_chunk`

Sender:
- case `LOCAL_FIN`:
   - case `LOCAL_COPIED`: complete sreq
   - case `REMOTE_COPIED`: wait for `RECV_FIN` to complete sreq
   - case `MIXED_COPIED`: wait for `RECV_FIN` to complete sreq
- case `ALL_FIN`:
   - case `REMOTE_COPIED`: complete sreq; send `FREE_CNT` with `cnt_ptr` to receiver
   - case `LOCAL_COPIED`: complete sreq; send `SEND_FIN` with `rreq_ptr` to receiver
   - case `MIXED_COPIED`: complete sreq; send `SEND_FIN` with `rreq_ptr` to receiver

Receiver:
- case `LOCAL_FIN`:
   - case `LOCAL_COPIED`: complete rreq; free cnt when receiving `FREE_CNT`
   - case `REMOTE_COPIED`: wait for `SEND_FIN` to complete rreq and free cnt
   - case `MIXED_COPIED`: wait for `SEND_FIN` to complete rreq and free cnt
- case `ALL_FIN`:
   - case `REMOTE_COPIED`: complete rreq and free cnt; do not send ACK since sender completes sreq after copy(`LOCAL_FIN && LOCAL_COPIED`)
   - case `LOCAL_COPIED`: complete rreq and free cnt; send `RECV_FIN` with `sreq_ptr` to sender
   - case `MIXED_COPIED`: complete rreq and free cnt; send `RECV_FIN` with `sreq_ptr` to sender

At finalize:
- The `shm_barrier` ensures that all processes should have completed the requests, otherwise the user program is not correct.
- Some counters might not be freed because the `FREE_CNT` packet may just be sent out and hasn't been seem by the receiver. But it is safely to release all counters now because the `shm_barrier` ensures completion on remote processes. If the process calls `free_cnt_cb` during finalize, it can skip `free(counter)` by checking a global flag `all_cnts_freed`.

What do you think? Any bug in the above logic? I just picked some random type/variable names, you can change it if you like.

Reply:
I agree with this design and logic is correct. I will implement this and push new changes soon.